File: python/onvo/resources/sessions.py
from onvo.base_resource import BaseResource


class Sessions(BaseResource):
    """Endpoints to get and update sessions. A session creates a link between a user under your organisation and a dashboard.

    Args:
        BaseResource (_type_): _description_
    """

    def list(self, dashboard_id: str):
        """List all sessions for a dashboard

        Args:
            dashboard_id (str): ID of the dashboard to list sessions for

        Returns:
            list: A list of all sessions for a dashboard
        """
        return self.base_get("/sessions", params={"parent_dashboard": dashboard_id})

    def revoke(self, dashboard_id: str):
        """Delete all sessions for a dashboard

        Args:
            dashboard_id (str): ID of the dashboard to delete all sessions for

        Returns:
            dict: A dictionary that shows status of deleting all sessions. Returns {"success": true} if successful.
        """
        return self.base_delete("/sessions", params={"parent_dashboard": dashboard_id})

    def upsert(self, dashboard_id: str, user_id: str):
        """If a session already exists, it is retrieved to save the user's changes and edits. If a session does not exist, one is created and is retrieved.

        Args:
            dashboard_id (str): ID of the dashboard to create a session for
            user_id (str): ID of the user to create a session for

        Returns:
            dict: A dictionary containing the details of the created/retrieved session
        """
        return self.base_post(
            "/sessions", data={"dashboard": dashboard_id, "user": user_id}
        )

    def delete(self, id):
        """Delete an existing session

        Args:
            id (str): The session id to delete

        Returns:
            dict: A dictionary that shows status of the delete. Returns {"success": true} if successful.
        """
        return self.base_delete(f"/sessions/{id}")

File: ruby/lib/onvo/sessions.rb
# frozen_string_literal: true

require_relative '../resource'

# Dashboard session endpoints
class Sessions < Resource
  # List all sessions for a dashboard
  #
  # @param dashboard_id [String]
  # @return [Array] Array of sessions
  def list(dashboard_id)
    base_get('/sessions', query: { parent_dashboard: dashboard_id })
  end

  # Revoke all sessions for a dashboard
  #
  # @param dashboard_id [String]
  # @return [Hash] status of delete. Returns {"success": true} if successful.
  def revoke(dashboard_id)
    base_delete('/sessions', query: { parent_dashboard: dashboard_id })
  end

  # TODO: Remove is not needed
  # # Delete an existing session
  # #
  # # @param id [String]
  # # @return [Hash] status of delete. Returns {"success": true} if successful.
  # def delete(id)
  #   base_delete("/sessions/#{id}")
  # end

  # Update or Create a session
  #
  # @param dashboard_id [String]
  # @param user_id [String] user id to create a session for
  # @return [Hash] details of session
  def upsert(dashboard_id, user_id)
    base_post(
      '/sessions',
      body: { dashboard: dashboard_id, user: user_id }
    )
  end
end
